types/logger/logger.go

```diff
@@ -18,13 +18,36 @@ import (
 	"strings"
 	"sync"
 	"time"
+
+	"context"
 )
 
 // Logf is the basic Tailscale logger type: a printf-like func.
 // Like log.Printf, the format need not end in a newline.
 // Logf functions must be safe for concurrent use.
 type Logf func(format string, args ...interface{})
 
+// A Context is a context.Context that should contain a custom log function, obtainable from FromContext.
+// If no log function is present, FromContext will return log.Printf.
+// To construct a Context, use Add
+type Context context.Context
+
+type logfKey struct{}
+
+// FromContext extracts a log function from ctx.
+func FromContext(ctx Context) Logf {
```

Comment:
so we log a line by writing `logger.FromContext(ctx).Logf("foo")`? seems a teensy bit wordy. no obvious alternatives in the typical go vocab, but maybe something like `logger.Printf(ctx, "foo")`?

Reply:
For sample usage see https://github.com/tailscale/corp/pull/3151

Reply:
I have had enough different opinions over the years that I'll leave this to someone else.